src/lib/x509/x509_ext.cpp

```diff
@@ -857,6 +863,841 @@ const std::string& TNAuthList::Entry::telephone_number() const {
+   for(IPAddressFamily& block : blocks) {
+      IPAddressChoice<V> choice = std::get<IPAddressChoice<V>>(block.addr_choice());
+      std::vector<IPAddressOrRange<V>> ranges = choice.ranges().value();
+      for(IPAddressOrRange<V>& r : ranges) {
+         merged_ranges.push_back(r);
+      }
+   }
+
+   // we have extracted all the ranges, and now rely on the constructor of IPAddressChoice to merge them
+   IPAddressChoice<V> choice(merged_ranges);
+   IPAddressFamily fam(choice, blocks[0].safi());
+   return fam;
+}
+
+namespace {
+
+constexpr auto IPv4 = IPAddressBlocks::Version::IPv4;
+constexpr auto IPv6 = IPAddressBlocks::Version::IPv6;
+
+template <IPAddressBlocks::Version V>
+using IPRangeVec = std::vector<IPAddressBlocks::IPAddressOrRange<V>>;
+
+// (S)AFI -> (needs_check, ptr to IPRangeVec)
+// the pointer can be null, in which case the boolean will be false, as such the pointer's value will never be looked at
+template <IPAddressBlocks::Version V>
+using IPValidationMap = std::map<uint32_t, std::pair<bool, const IPRangeVec<V>*>>;
+
+template <typename T>
+std::optional<std::vector<T>> sort_and_merge_ranges(std::optional<std::span<const T>> ranges) {
+   // This method iteratively both sorts and combines IPAddressOrRange or ASIdOrRange objects.
+   // cf. https://www.rfc-editor.org/rfc/rfc3779.html#section-2.2.3.6 and https://www.rfc-editor.org/rfc/rfc3779.html#section-3.2.3.4
+   // This implementation uses only min-max ranges internally, so sorting by the prefix length is not necessary / impossible here.
+   //
+   // We first sort all range objects by their minimum value, then check if two adjacent elements can be combined into one
+   // (either one ends and the other immediately starts, or they overlap), in which case the two are removed, and a new combined object is added.
+   // This process is repeated until there are no further changes, because multiple adjacent elements may need to be combined one by one.
+
+   if(!ranges.has_value()) {
+      return std::nullopt;
+   }
+
+   std::vector<T> current(ranges.value().begin(), ranges.value().end());
+
+   if(current.empty()) {
+      return current;
+   }
+   bool changed = true;
+   while(changed) {
+      changed = false;
+      // sort by size of the min value
+      std::sort(current.begin(), current.end(), [](T& a, T& b) { return a.min() < b.min(); });
+      for(size_t i = 0; i < current.size() - 1; i++) {
+         const T a = current[i];
+         const T b = current[i + 1];
+         // they either overlap or are adjacent
+         if(b.min() <= a.max() || b.min() == (a.max() + 1)) {
+            // erase old a and b
+            current.erase(current.begin() + i, current.begin() + i + 2);
+            current.insert(current.begin() + i, T(a.min(), std::max(a.max(), b.max())));
+            // look again at the new element we just created, might need to be merged with others still
+            i -= 1;
+            changed = true;
+         }
+      }
+   }
```

Comment:
The while-loop is kind of redundant, at least with the `i -= 1` present. The latter in combination with the sort ensures that all possible merges happen, so the loop will always terminate at latest in the second iteration. The `i -= 1` method does lead to a nice size_t underflow if we merge on the first iteration, which is then directly incremented back to 0 again by the for-loop, but I think we should decide on one or the other, not both.

Reply:
To me the while loop seems easier to understand

Re the for loop, if retained please do leave a comment calling out the underflow when `i = 0` iiuc it's safe but it's a great example of something subtle that someone might "rediscover" years down the line and then spend time trying to figure out if it's a problem or not.